optuna/samplers/nsgaiii.py

```diff
@@ -0,0 +1,542 @@
+from collections import defaultdict
+import hashlib
+import itertools
+import math
+from typing import Any
+from typing import Callable
+from typing import DefaultDict
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Tuple
+import warnings
+
+import numpy as np
+
+import optuna
+from optuna.distributions import BaseDistribution
+from optuna.exceptions import ExperimentalWarning
+from optuna.samplers._base import _CONSTRAINTS_KEY
+from optuna.samplers._base import _process_constraints_after_trial
+from optuna.samplers._base import BaseSampler
+from optuna.samplers._random import RandomSampler
+from optuna.samplers._search_space import IntersectionSearchSpace
+from optuna.samplers.nsgaii._crossover import perform_crossover
+from optuna.samplers.nsgaii._crossovers._base import BaseCrossover
+from optuna.samplers.nsgaii._crossovers._uniform import UniformCrossover
+from optuna.samplers.nsgaii._sampler import _constrained_dominates
+from optuna.study import Study
+from optuna.study._multi_objective import _dominates
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+# Define key names of `Trial.system_attrs`.
+_GENERATION_KEY = "nsga3:generation"
+_POPULATION_CACHE_KEY_PREFIX = "nsga3:population"
+
+
+class NSGAIIISampler(BaseSampler):
+    """Multi-objective sampler using the NSGA-III algorithm.
+
+    NSGA-III stands for "Nondominated Sorting Genetic Algorithm III",
+    which is a modified version of NSGA-II for many objective optimization problem.
+
+    For further information about NSGA-III, please refer to the following papers:
+
+    - `An Evolutionary Many-Objective Optimization Algorithm Using Reference-Point-Based
+    Nondominated Sorting Approach, Part I: Solving Problems With Box Constraints
+    <https://ieeexplore.ieee.org/document/6600851>`_
+    - `An Evolutionary Many-Objective Optimization Algorithm Using Reference-Point-Based
+    Nondominated Sorting Approach, Part II: Handling Constraints and Extending to an Adaptive
+    Approach <https://ieeexplore.ieee.org/document/6595567>`_
+
+    Args:
+        reference_points:
+            A 2 dimension ``numpy.ndarray`` with objective dimension columns. Represents
+            a list of reference points which is used to determine who to survive.
+            After non-dominated sort, who out of borderline front are going to survived is
+            determined according to how sparse the closest reference point of each person is.
+            In the default setting the algorithm uses `uniformly` spread points to diversify the
+            result.
+
+            .. note::
+                Other parameters are the same as :class:`~optuna.samplers.nsgaii.NSGAIISampler`.
+
+
+    """
```

Comment:
Maybe we have too many blank lines.
```suggestion
    """
```